### PR TITLE
Add DistributingDownstream BatchConsumer implementation

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/distributed/DistributingConsumer.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                    public void onResponse(DistributedResultResponse response) {
+                        downstream.needsMoreData = response.needMore();
+                        countdownAndMaybeContinue(it, numActiveRequests);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        failure = e;
+                        downstream.needsMoreData = false;
+                        countdownAndMaybeContinue(it, numActiveRequests);
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/distributed/DistributingConsumer.java]: imho we must forward the failure if need, but not continue consuming...

File: sql/src/main/java/io/crate/operation/projectors/ColumnIndexWriterProjector.java [hunk 2]
```diff
@@ -148,18 +141,24 @@ public Result setNextRow(Row row) {
 
     @Override
     public void finish(RepeatHandle repeatHandle) {
+        BulkProcessorFutureCallback bulkProcessorFutureCallback = new BulkProcessorFutureCallback(failed, downstream);
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/projectors/ColumnIndexWriterProjector.java]: are these changes related to the commits?

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterProjector.java [hunk 3]
```diff
@@ -145,11 +138,15 @@ public Result setNextRow(Row row) {
 
     @Override
     public void finish(RepeatHandle repeatHandle) {
+        BulkProcessorFutureCallback bulkProcessorFutureCallback = new BulkProcessorFutureCallback(failed, downstream);
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterProjector.java]: also in this file, are they related to the commits? if not, make a separate commit pls.

File: sql/src/test/java/io/crate/executor/transport/distributed/DistributingConsumerTest.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        assertThat(TestingHelpers.printedTable(new CollectionBucket(result)),
+            is("0\n" +
+               "1\n" +
+               "2\n" +
+               "3\n" +
+               "4\n"));
+
+        // pageSize=2 and 5 rows causes 3x pushResult
+        verify(distributedResultAction, times(3)).pushResult(anyString(), any(), any());
+    }
```
Comment (seut) [inline sql/src/test/java/io/crate/executor/transport/distributed/DistributingConsumerTest.java]: regressions are already tested due to DistributingDownstream tests?

File: sql/src/main/java/io/crate/operation/projectors/ColumnIndexWriterProjector.java (hunk 2 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/projectors/ColumnIndexWriterProjector.java]: they're related as currently we still create the projector and then switch to the Consumer/BatchIterator pattern if possible. 
The problem with the callback was that this was still triggered even if the switch to the consumer was made - resulting in both the RowReciever & the Consumer being used.

File: sql/src/test/java/io/crate/executor/transport/distributed/DistributingConsumerTest.java (hunk 4 above)
Comment (mfussenegger) [inline sql/src/test/java/io/crate/executor/transport/distributed/DistributingConsumerTest.java]: There are some tests in DistributingDownstreamTest which no longer make sense (pause handling), but some which probably will need to be ported. Do you think it's okay to do so in the PR/commit that will remove the existing DistributingDownstream?

Comment (seut) [inline sql/src/test/java/io/crate/executor/transport/distributed/DistributingConsumerTest.java]: yes would be ok. I've ask more about the forwarding failure stuff at the DistributingConsumer, how that is currently tested..
